### PR TITLE
`std.atomic`: `cache_line` and `spinLoopHint()` specialization for more architectures

### PR DESCRIPTION
**Pull request by alexrp**

### CODE REVIEW

File: lib/std/atomic.zig [hunk 1]
```diff
@@ -402,6 +414,22 @@ pub inline fn spinLoopHint() void {
                 asm volatile ("" ::: "memory");
             }
         },
+
+        // The 8-bit immediate specifies the amount of cycles to pause for. We can't really be too
+        // opinionated here.
+        .hexagon,
+        => asm volatile ("pause(#1)" ::: "memory"),
```
Comment (alexrp) [inline lib/std/atomic.zig:421]: I'm a bit on the fence with this one. Unlike the others, it is not merely a hint, but actually *will* suspend execution for a number of cycles (1 in this case). It is still explicitly mentioned as appropriate for spin locks, however. See §8.9 here: https://developer.qualcomm.com/qfile/70668/80-n2040-53_aa_prm_v73.pdf

Comment (andrewrk) [reply lib/std/atomic.zig:421]: I think this is a valid lowering of the documented semantics ("Signals to the processor that the caller is inside a busy-wait spin-loop.") :+1:

**Review by andrewrk — APPROVED** [review]

**Comment by andrewrk** [comment]
By the way, why do these all say that they clobber memory? I don't think that's actually true.

**Comment by alexrp** [comment]
Honestly, no idea. The [fallback](https://github.com/ziglang/zig/blob/e1c965b23db2700c2641dffd97afd183eac37aec/lib/std/atomic.zig#L435) even explicitly does it to impose an optimization barrier. But why? This is a hint; it's not supposed to have semantics like this. Seems like `volatile` should be enough here.

I can remove these clobbers if you agree that doing so should be safe.

**Comment by andrewrk** [comment]
I think I know why that is there. The author intends the spin lock to happen at a specific point in between memory loads and stores, so lying to the compiler and telling it that it can clobber memory causes the spin lock to be fixed in place, not adjusted before or after any memory accesses.

Of course, in Zig land, we don't make components lie to each other to manipulate outcomes. We tell the truth in excruciating detail so that all components can behave correctly. That means a memory clobber is the wrong tool for the job here. The right tool is of course atomic orderings (not `volatile` unless memory-mapped I/O is involved).

In conclusion, yes the memory clobbers should be removed because code that interacts with spin locks will already be using atomics to protect memory accesses.

**Comment by alexrp** [comment]
> The right tool is of course atomic orderings (not volatile unless memory-mapped I/O is involved).

Sorry, to be clear, I meant `asm volatile` so that the compiler doesn't optimize it away because it has no explicit I/O. 🙂

**Comment by andrewrk** [comment]
Ah, yes of course :-)
